include/alpaka/core/BindScope.hpp

```diff
@@ -0,0 +1,231 @@
+        };
+
+        // HostAndDevice HIP/HCC: explicitly define host+device context
+        template<typename TFunc>
+        struct BindScope<TFunc, Scope::HostAndDevice> {
+            const TFunc m_callable;
+
+            ALPAKA_FN_HOST_ACC
+            explicit BindScope(const TFunc callable) : m_callable{callable} {}
+
+            template<typename... TArgs>
+            ALPAKA_FN_HOST_ACC
+            auto operator()(TArgs&&... args) const -> decltype(m_callable(std::forward<TArgs>(args)...)) {
+                return m_callable(std::forward<TArgs>(args)...);
+            }
+
+            ALPAKA_FN_HOST_ACC ~BindScope(){} // destructor must reflect host-device annotation of constructors
+        };
+
+        // HostOnly HIP/HCC:
+        template<typename TFunc>
+        struct BindScope<TFunc, Scope::HostOnly> {
+            const TFunc m_callable;
+
+            ALPAKA_FN_HOST_ACC
+            explicit BindScope(const TFunc callable) : m_callable{callable} {}
+
+            template<typename... TArgs>
+            ALPAKA_FN_HOST
+            auto operator()(TArgs&&... args) const -> decltype(m_callable(std::forward<TArgs>(args)...)) {
+                return m_callable(std::forward<TArgs>(args)...);
+            }
+            // do not use decltype for trailing return type, it will break HCC with AMP restricted call in host function
+            template<typename T, typename... TArgs>
+            __device__
+            auto operator()(T&&, TArgs&&... args) const -> void { // void, otherwise callable() must be invoked for decltype()
+                ALPAKA_ASSERT(false);
+            }
+
+            ALPAKA_FN_HOST_ACC ~BindScope(){} // destructor must reflect host-device annotation of constructors
+        };
+
+        template<typename TFunc>
+        struct BindScope<TFunc, Scope::DeviceOnly> {
+            const TFunc m_callable;
+
+            ALPAKA_FN_HOST_ACC
+            explicit BindScope(const TFunc callable) : m_callable{callable} {}
+
+            template<typename... TArgs>
+            ALPAKA_FN_HOST
+            auto operator()(TArgs&&... args) const -> void { // void, otherwise callable() must be invoked for decltype()
+                ALPAKA_ASSERT(false);
+            }
+            template<typename... TArgs>
+            __device__
+            auto operator()(TArgs&&... args) const -> decltype(m_callable(std::forward<TArgs>(args)...)) {
+                return m_callable(std::forward<TArgs>(args)...);
+            }
+
+            ALPAKA_FN_HOST_ACC ~BindScope(){} // destructor must reflect host-device annotation of constructors
+        };
+#endif // HCC
+
+    } // core
+} // alpaka
+
+// macro definitions to provide shortcuts
+#define ALPAKA_FN_SCOPE_HOST alpaka::core::bindScope< alpaka::core::Scope::HostOnly >
+#define ALPAKA_FN_SCOPE_DEVICE alpaka::core::bindScope< alpaka::core::Scope::DeviceOnly >
+#define ALPAKA_FN_SCOPE_HOST_AND_DEVICE alpaka::core::bindScope< alpaka::core::Scope::HostAndDevice >
```

Comment:
Do we need the explicit `AND` in this macros name?

Reply:
no